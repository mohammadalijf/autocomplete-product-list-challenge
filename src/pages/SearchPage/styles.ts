import { makeStyles } from "@material-ui/core";

const useSearchPageStyles = makeStyles((theme) => ({
  container: {
    padding: theme.spacing(2),
    display: "flex",
    flexDirection: "column",
    "& > * ": {
      marginTop: theme.spacing(3),
    },
    "& > *:first-child": {
      marginTop: 0,
    },
  },
  searchForm: {
    display: "grid",
    gridGap: theme.spacing(2),
    gridTemplateAreas: `
    "searchInput"
    "genderInput"
    "onSaleInput"`,
    [theme.breakpoints.up("sm")]: {
      gridTemplateAreas: `
      "searchInput searchInput genderInput"
      "onSaleInput onSaleInput onSaleInput"`,
    },
    [theme.breakpoints.up("md")]: {
      gridTemplateAreas: `
      "searchInput searchInput searchInput genderInput onSaleInput"`,
    },
  },
  searchInput: {
    gridArea: "searchInput",
  },
  genderInput: {
    gridArea: "genderInput",
  },
  saleInput: {
    gridArea: "onSaleInput",
  },
}));

export default useSearchPageStyles;
