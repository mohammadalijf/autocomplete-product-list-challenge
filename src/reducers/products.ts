import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import productAPI, {
  CSVProductGender,
  ICSVProducts,
} from "../services/productsAPI";

export class NotFoundProductError extends Error {
  constructor() {
    super("Sorry! we didn't find anything... maybe try for another product?");
    this.name = "NotFoundProductError";
  }
}

export interface IProductsState {
  products: ICSVProducts[];
  filtered: ICSVProducts[];
  loading: {
    products: boolean;
    filtered: boolean;
  };
  error: {
    products?: string;
    filtered?: string;
  };
}

/**
 * fetch thunk action for getting products from csv
 */
export const fetchProducts = createAsyncThunk("products/fetch", async () => {
  return await productAPI.fetchProducts();
});

/**
 * filter thunk action for products
 */
export const searchProducts = createAsyncThunk<
  ICSVProducts[],
  {
    query?: string;
    gender?: CSVProductGender;
    onSale?: boolean;
  }
>("products/filter", async (args, thunkAPI) => {
  const { gender, onSale, query } = args;
  if (!query) return [];
  const { products } = (thunkAPI.getState() as { products: IProductsState })
    .products;
  if (products.length === 0) return [];

  const filtered = await productAPI.searchProducts(
    products,
    { query, gender, onSale },
    thunkAPI.signal
  );
  if (filtered.length === 0) {
    throw new NotFoundProductError();
  }
  return filtered;
});

const productSlice = createSlice({
  name: "products",
  initialState: {
    products: [],
    filtered: [],
    loading: { products: false, filtered: false },
    error: {},
  } as IProductsState,
  reducers: {},
  extraReducers: (builder) => {
    builder.addCase(fetchProducts.pending, (state) => {
      state.loading.products = true;
      state.error.products = undefined;
      state.products = [];
    });
    builder.addCase(fetchProducts.fulfilled, (state, action) => {
      state.loading.products = false;
      state.products = action.payload;
    });
    builder.addCase(fetchProducts.rejected, (state, action) => {
      state.loading.products = false;
      state.error.products = action.error.message;
    });
    builder.addCase(searchProducts.pending, (state) => {
      state.error.filtered = undefined;
      state.loading.filtered = true;
      state.filtered = [];
    });
    builder.addCase(searchProducts.fulfilled, (state, action) => {
      state.filtered = action.payload;
      state.loading.filtered = false;
    });
    builder.addCase(searchProducts.rejected, (state, action) => {
      state.error.filtered = action.error.message;
      state.loading.filtered = false;
    });
  },
});

export default productSlice.reducer;
