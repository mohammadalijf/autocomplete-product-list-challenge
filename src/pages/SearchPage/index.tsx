import { Container, debounce, Typography } from "@material-ui/core";
import { Pagination, PaginationProps } from "@material-ui/lab";
import {
  FormEventHandler,
  FunctionComponent,
  useCallback,
  useEffect,
  useMemo,
  useState,
} from "react";
import GenderInput, {
  GenderInputChangeEvent,
  GenderInputOption,
} from "../../components/GenderInput";
import OnSaleSwitch from "../../components/OnSaleSwitch";
import { OnSaleSwitchChangeEvent } from "../../components/OnSaleSwitch/props";
import SearchInput, {
  SearchInputChangeEvent,
} from "../../components/SearchInput";
import { useAppDispatch, useAppSelector } from "../../hooks/redux";
import { searchProducts } from "../../reducers/products";
import SearchProductGrid from "./components/SearchProductGrid";
import useSearchPageStyles from "./styles";

/**
 * Container for searching products
 */
const SearchPage: FunctionComponent = (props) => {
  const classes = useSearchPageStyles();
  const dispatch = useAppDispatch();

  /**
   * Debounced Function for setting query state
   * to avoid unnecessary filter actions dispatch
   */
  const debounced = useMemo(() => {
    return debounce(
      (
        setQuery: React.Dispatch<React.SetStateAction<string>>,
        query: string
      ) => {
        setQuery(query);
      },
      500
    );
  }, []);

  //#region store selectors

  const filteredProducts = useAppSelector((state) => state.products.filtered);
  const loading = useAppSelector((state) => state.products.loading.filtered);
  const error = useAppSelector((state) => state.products.error.filtered);

  //#endregion

  //#region states

  const [query, setQuery] = useState("");
  const [searchInputValue, setSearchInputValue] = useState("");
  const [gender, setGender] = useState<GenderInputOption | undefined>();
  const [onlyOnSales, setOnlyOnSales] = useState(false);
  const [page, setPage] = useState(1);

  //#endregion

  //#region handlers

  /**
   * Change handler for SearchInput
   */
  const queryChangeHandler: SearchInputChangeEvent = useCallback(
    (query) => {
      setSearchInputValue(query);
      debounced(setQuery, query);
    },
    [debounced, setSearchInputValue, setQuery]
  );

  /**
   * Change handler for GenderInput
   */
  const genderChangeHandler: GenderInputChangeEvent = useCallback(
    (option) => {
      setGender(option);
    },
    [setGender]
  );

  /**
   * Change handler for OnSaleSwitch
   */
  const onSaleSwitchChangeHandler: OnSaleSwitchChangeEvent = useCallback(
    (checked) => {
      setOnlyOnSales(checked);
    },
    [setOnlyOnSales]
  );

  /**
   * Submit handler for search form
   */
  const formSubmitHandler: FormEventHandler<HTMLFormElement> = (event) => {
    event.preventDefault();
  };

  const paginationChangeHandler: PaginationProps["onChange"] = (
    event,
    value
  ) => {
    setPage(value);
  };

  //#endregion

  //#region side effects

  useEffect(() => {
    // dispatch filter action when queries are changed
    const promise = dispatch(
      searchProducts({
        query,
        onSale: onlyOnSales,
        gender,
      })
    );
    return () => {
      promise.abort();
    };
  }, [query, gender, onlyOnSales, dispatch]);

  useEffect(() => {
    // set page to 1 when products change
    setPage(1);
  }, [filteredProducts.length, setPage]);

  useEffect(() => {
    // scroll to top when page is changed
    window.scrollTo({
      top: 0,
    });
  }, [page]);

  const { showInstructions, pageCount } = useMemo(() => {
    return {
      // should show search instructions
      showInstructions:
        !loading && query === "" && filteredProducts.length === 0,
      // calculate page count
      pageCount: Math.ceil(filteredProducts.length / 100),
    };
  }, [filteredProducts.length, query, loading]);

  const paginatedProducts = useMemo(() => {
    return filteredProducts.slice((page - 1) * 100, 100 * page);
  }, [filteredProducts, page]);

  //#endregion

  return (
    <Container className={classes.container}>
      <Typography variant="h3" component="h1">
        Autocomplete Product List Challenge
      </Typography>
      <form
        className={classes.searchForm}
        action="."
        onSubmit={formSubmitHandler}
      >
        <SearchInput
          autoFocus
          className={classes.searchInput}
          value={searchInputValue}
          onChange={queryChangeHandler}
        />
        <GenderInput
          className={classes.genderInput}
          value={gender}
          onChange={genderChangeHandler}
        />
        <OnSaleSwitch
          className={classes.genderInput}
          checked={onlyOnSales}
          onChange={onSaleSwitchChangeHandler}
        />
      </form>
      <SearchProductGrid
        loading={loading}
        error={error}
        showInstructions={showInstructions}
        products={paginatedProducts}
      />
      {pageCount > 1 && (
        <Pagination
          data-testid="productPagination"
          className={classes.pagination}
          color="primary"
          count={pageCount}
          page={page}
          onChange={paginationChangeHandler}
        />
      )}
    </Container>
  );
};

export default SearchPage;
