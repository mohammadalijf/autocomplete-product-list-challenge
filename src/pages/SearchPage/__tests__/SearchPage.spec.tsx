import {
  cleanup,
  fireEvent,
  render,
  RenderResult,
} from "@testing-library/react";
import { act, createRenderer } from "react-dom/test-utils";
import { Provider } from "react-redux";
import SearchPage from "..";
import { fetchProducts } from "../../../reducers/products";
import { store } from "../../../store";

describe("<SearchPage />", () => {
  afterEach(cleanup);
  window.scrollTo = () => {};
  let searchPage: RenderResult;

  beforeEach(async () => {
    await store.dispatch(fetchProducts());
    searchPage = await render(
      <Provider store={store}>
        <SearchPage />
      </Provider>
    );
  });

  it("renders correctly", () => {
    expect(searchPage.asFragment()).toMatchSnapshot();
  });

  it("should show correct information", async () => {
    const searchInput = searchPage.getByTestId("searchInput");
    searchInput.focus();
    await act(async () => {
      fireEvent.change(searchPage.getByTestId("searchInput"), {
        target: { value: "slim" },
      });
      await new Promise((r) => setTimeout(r, 500));
    });

    expect(searchPage.queryAllByTestId("productCell")).toHaveLength(100);
    expect(
      searchPage.getByTestId("productPagination").getElementsByTagName("li")
    ).toHaveLength(5);
  });
});
