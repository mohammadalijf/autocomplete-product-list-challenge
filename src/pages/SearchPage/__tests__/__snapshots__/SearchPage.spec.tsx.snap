// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchPage /> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root makeStyles-container-1 MuiContainer-maxWidthLg"
  >
    <h1
      class="MuiTypography-root MuiTypography-h3"
    >
      Autocomplete Product List Challenge
    </h1>
    <form
      action="."
      class="makeStyles-searchForm-2"
    >
      <div
        class="MuiFormControl-root MuiTextField-root makeStyles-searchInput-3"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-focused Mui-focused"
          data-shrink="true"
        >
          Search
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root Mui-focused Mui-focused MuiInputBase-formControl"
        >
          <input
            aria-invalid="false"
            autocomplete="products"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputTypeSearch"
            data-testid="searchInput"
            type="search"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-9 PrivateNotchedOutline-legendNotched-10"
            >
              <span>
                Search
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
          data-shrink="false"
          id="gender-input-label"
        >
          Gender
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root makeStyles-genderInput-4 MuiInputBase-formControl"
        >
          <select
            aria-invalid="false"
            class="MuiSelect-root MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input"
            data-testid="genderInput"
            id="gender-input"
          >
            <option
              aria-label="None"
            />
            <option
              value="unisex"
            >
              Unisex
            </option>
            <option
              value="female"
            >
              Female
            </option>
            <option
              value="male"
            >
              Male
            </option>
          </select>
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PrivateNotchedOutline-root-7 MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="PrivateNotchedOutline-legendLabelled-9"
            >
              <span>
                Gender
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <label
        class="MuiFormControlLabel-root"
      >
        <span
          class="MuiSwitch-root makeStyles-genderInput-4"
        >
          <span
            aria-disabled="false"
            class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-11 MuiSwitch-switchBase MuiSwitch-colorPrimary"
            data-testid="onSaleSwitch"
          >
            <span
              class="MuiIconButton-label"
            >
              <input
                aria-label="show only on sales products"
                class="PrivateSwitchBase-input-14 MuiSwitch-input"
                type="checkbox"
                value=""
              />
              <span
                class="MuiSwitch-thumb"
              />
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </span>
          <span
            class="MuiSwitch-track"
          />
        </span>
        <span
          class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
        >
          Show only on sales
        </span>
      </label>
    </form>
    <section
      class="makeStyles-container-15 makeStyles-centered-16"
      data-testid="searchProductGrid"
    >
      <p
        class="MuiTypography-root makeStyles-instructionsLabel-17 MuiTypography-body1"
      >
        Start searching for products by typing in the search box
      </p>
    </section>
  </div>
</DocumentFragment>
`;
